f5/bigip/__init__.py

```diff
@@ -65,6 +66,25 @@ def hostname(self):
     def icontrol_version(self):
         return self._meta_data['icontrol_version']
 
+    @property
+    def tmos_version(self):
+        # We return tmos_version, or if None,
+        # we grab it and return value via _get_tmos_version(),
+        # at the same time updating _meta_data['tmos_version']
+        return self._meta_data['tmos_version'] or self._get_tmos_version()
+
+    def _get_tmos_version(self):
+        connect = self._meta_data['bigip']._meta_data['icr_session']
+        # Need to add this due to BIGIP class URI manipulation
+        if self._meta_data['uri'].endswith('tm/'):
+            base_uri = self._meta_data['uri'] + 'sys/'
+        else:
+            base_uri = self._meta_data['uri'] + 'tm/sys/'
```

Comment:
Unless the inheritance mechanisms changed, how is the method  _get_tmos_version in BigIP be inherited by ManagementRoot? If one uses correct syntax i.e mr = ManagementRoot('host', 'user', 'pass) this will break horribly.

One way around it, is to use my previous method where I utilized SDK to obtain device's version. This way we will not have any self link parsing etc, but we will have to bypass tmos version check first time (as you remember my previous method looked like this):
````
    @property
    def version(self):
        mgmt_rt = self
        dvcs = mgmt_rt.tm.cm.devices.get_collection()
        for i in range(len(dvcs)):
            if str(dvcs[i].selfDevice) == 'true':
                ver = str(dvcs[i].version)
                self._meta_data['tmos_version'] = ver
````

Hence my earlier comment:

> @zancas: I do not necessarily agree with the below:

    Remove the check for the existence of the tmos_version key. We should GUARANTEE the existence of that key as part of the interface to ManagementRoot. I.E. no "if tmos_v: instantiation of ManagementRoot

The other way will be to specify a 'uri' method, which is really just breaking up the base_uri logic into 2 methods for each class, for me it will be just moving code around which does not change anything.